### PR TITLE
Hotfix/17512 duplication of wp with invalid watcher

### PR DESCRIPTION
**Pull request by hschink**
[`* `#17512` Duplication of work package with invalid watcher`](https://community.openproject.org/work_packages/17512)

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage. <a href='http://hotfix-17512-duplication-of-wp-with-invalid-watcher.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

File: lib/services/create_watcher.rb [hunk 1]
```diff
@@ -42,7 +42,7 @@ def run(success = -> {}, failure = -> {})
         @work_package.watchers << @watcher
         success.(created: true)
       else
-        error.(@watcher)
+        failure.(@watcher)
```
Comment (myabc) [inline lib/services/create_watcher.rb:45]: @hschink good catch

File: lib/plugins/acts_as_watchable/lib/acts_as_watchable.rb [hunk 2]
```diff
@@ -89,10 +89,10 @@ def possible_watcher_users
           # TODO: There might be addable users which are not in the current
           #       project. But its hard (performance wise) to find them
           #       correctly. So we only search for users in the project scope.
-          #       Also, the watchable might not be allowed to be seen by all
-          #       users in the project.
           #       This implementation is so wrong, it makes me sad.
-          project.users
+          watchers = project.users
+          watchers = watchers.select { |w| visible?(w) } if respond_to?(:visible?)
```
Comment (myabc) [inline lib/plugins/acts_as_watchable/lib/acts_as_watchable.rb:94]: @hschink shouldn't `respond_to?(:visible?)` be inside of the `select` block?

Comment (hschink) [inline lib/plugins/acts_as_watchable/lib/acts_as_watchable.rb:94]: Taking your question into account this statement would then become something like the following:

``` Ruby
watchers = watchers.select { |w| respond_to?(:visible?) ? visible?(w) : true }
```

With the `if` statement at the end I prevent the execution of the `select` which should have better performance in the case where `visible?` does not exist.

Comment (myabc) [inline lib/plugins/acts_as_watchable/lib/acts_as_watchable.rb:94]: @hschink Mea culpa. Yes, you're completely right. I misread this.

**Pull request merged**